[JENKINS-39300] View.getDisplayName() is ignored when rendering tabs

stephenc opened the pull request:
See [JENKINS-39300](https://issues.jenkins-ci.org/browse/JENKINS-39300)

Also:
- adds support for descriptor visibility filtering on views
- fixes l10n of AllView
- adds support for contextual create permission of specific view types
- some other small code fixes spotted while touching the files

@jenkinsci/code-reviewers @reviewbybees

stephenc committed:
[FIXED JENKINS-39300] Use View.getDisplayName() when rendering the view tabs

- Also fix the stapler documentation for the <l:tab> tag

stephenc committed:
[JENKINS-39300] Add displayName validation API to ViewDescriptor

stephenc committed:
[JENKINS-39300 Follow-up] Fix the AllView tab title to reflect l10n

- Also while we are at it, modernise ViewDescriptor instantiation to allow for descriptor visibility filtering

stephenc committed:
[JENKINS-39300 Follow-up] Allow create permission restriction for types of views

stephenc committed:
Redundant null check removed

stephenc commented on core/src/main/java/hudson/model/View.java:
```diff
@@ -1219,11 +1219,20 @@ public ContextMenu doChildrenContextMenu(StaplerRequest request, StaplerResponse
+     * Returns the {@link ViewDescriptor} instances that can be instantiated for the {@link ViewGroup} in the current
+     * {@link StaplerRequest}. <strong>NOTE: This method must be called from a {@link StaplerRequest}</strong>
+     * @return the list of instantiable {@link ViewDescriptor} instances for the current {@link StaplerRequest}
+     */
```
Note: this is the javadoc that applied to the method *before* my changes as the method would have blown up with a NPE if invoked outside of the `newView.jelly` screen of a `ViewGroup` as soon as `AllView.isInstantiable()` was called

jglick replied:
Perhaps just check `owner == null` and return `all()`?

jglick replied:
Need to also check whether the `StaplerRequest` is null.

stephenc committed:
[JENKINS-39300 Follow-up] JobFilters and Columns should work with DescriptorVisiblilityFilters too

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

stephenc committed:
[JENKINS-39300 Follow-up] Default columns should respect DescriptorVisiblilityFilters

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3052,7 +3052,7 @@ public void run(Reactor session) throws Exception {
                 // this is both for clean Jenkins and for backward compatibility.
                 if(views.size()==0 || primaryView==null) {
-                    View v = new AllView(Messages.Hudson_ViewName());
+                    View v = new AllView("all");
```
This looks like it fixes JENKINS-38606.

oleg-nenashev replied:
It is. But will it cause a fallout in upgrading Jenkins instances?

stephenc replied:
Will only affect new instances. Old instances can get the fix by recreating the All view (add new view, change default, delete all view, add back all view, change default, tidy up)

daniel-beck replied:
This looks like a reasonable default for new All views. No need to edit existing ones. Would just cause confusion and broken links.

daniel-beck commented on core/src/main/resources/lib/layout/tab.jelly:
```diff
@@ -22,24 +22,32 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+      Whether the tab is active or not
+    </st:attribute>
+    <st:attribute name="title">
+      The title of the tab
```
Do we have an `@since` equivalent for Jelly tags?

stephenc replied:
I am restoring behaviour that was broken in https://github.com/stephenc/jenkins/commit/2de93f861a8e6997d8f4b7b186b7953c835771cc#diff-1948650f90f2cefd48aff59fa5742aa0L58 (appears to be removed by omission as https://github.com/stephenc/jenkins/blob/2de93f861a8e6997d8f4b7b186b7953c835771cc/core/src/main/resources/lib/layout/tab.jelly#L29 was retained

stephenc replied:
https://github.com/stapler/stapler/pull/84 created for you ;-)

stephenc committed:
[FIXED JENKINS-38606] Detect and fix localized AllView name if and only if safe to fix

- Don't fix if there is already a view with the new default name
- Don't fix if the AllView is not the primary view

stephenc commented:
@daniel-beck @oleg-nenashev so I have added a fix for JENKINS-38606... and we need to be able to invoke that fix from plugins (e.g. folders plugin) so I cannot make that `@Restricted` before you ask!

daniel-beck commented:
> have added a fix for JENKINS-38606

Not actually safe, as there are code paths that add the view name to the URL even if it's the default view.

I don't think we should fix JENKINS-38606 for existing installations. It modifies user data for a pretty weak reason.

stephenc committed:
[JENKINS-38606] Typos

stephenc committed:
[JENKINS-38606] Require the user to Opt-in to the fix

stephenc commented:
@daniel-beck OK, I have made this opt-in. I still think we should have the ability for an admin to force the fix applied everywhere as it could be quite difficult to fix all the `AllView`s scattered everywhere otherwise

daniel-beck commented:
Is this really not trivial to take care of in a script console script with maybe five lines?

Do we really need a startup parameter (to be added in service configurations) for a one-time cleanup action here?

stephenc commented:
@daniel-beck really? show me the five lines that find *every* ViewGroup (including the User properties for *all* users) and the folders etc and does the fix-up? For most admins it should not be an issue... but if it is an issue it could be very painful to fix by script or manually. Much better to provide a hook to fix the issue that is available to everyone IMHO.

stephenc commented:
@daniel-beck can you respond please? Would like to get the change at least merged before we hit one of the last likely candidates for the next LTS

daniel-beck commented:
I still have my doubt but I'd be happy to be outvoted. So far, nobody else has cared enough though  @jenkinsci/code-reviewers Ping!). FWIW I would like to see other opinions on solving this using a new parameter.

rsandell commented:
:bee:

stephenc committed:
[JENKINS-39300] Sort the views by displayName

jglick commented on core/src/main/java/hudson/model/AllView.java:
```diff
@@ -89,13 +116,70 @@ protected void submit(StaplerRequest req) throws IOException, ServletException,
+     * @since FIXME
+     */
+    @Nonnull
+    public static String applyJenkins38606Fixup(@Nonnull List<View> views, @Nonnull String primaryView) {
```
```java
@Restricted(NoExternalUse.class)
```

(and remove `@since`) unless this is something you really expect plugins to use.

stephenc replied:
I expect plugins that implement ViewGroup will need to call into this method

jglick replied:
Is there a proposed PR calling it? From `cloudbees-folder`? Then you should probably have a clearer name.

jglick commented on core/src/main/java/hudson/model/AllView.java:
```diff
@@ -89,13 +116,70 @@ protected void submit(StaplerRequest req) throws IOException, ServletException,
+     * fixing the view name by setting the system property {@code hudson.mode.AllView.JENKINS-38606} to {@code true}.
+     * Use this method to round-trip the primary view name, e.g.
+     * {@code primaryView = applyJenkins38606Fixup(views, primaryView)}
+     * NOTE: we can only fix the localized name of an {@link AllView} if it is the primary view as otherwise urls
```
```html
<p>
```

jglick commented on core/src/main/java/hudson/model/AllView.java:
```diff
@@ -89,13 +116,70 @@ protected void submit(StaplerRequest req) throws IOException, ServletException,
+     * {@code primaryView = applyJenkins38606Fixup(views, primaryView)}
+     * NOTE: we can only fix the localized name of an {@link AllView} if it is the primary view as otherwise urls
+     * would change, whereas the primary view is special and does not normally get accessed by the
+     * {@code /view/_name_} url.
```
…though it _can_ be, and those URLs will break. Right?

daniel-beck replied:
Yes, and there are links in Jenkins to the default view's `/view/whatever` URL.

amuniz replied:
Yes, actually the new item link has switched from `/view/All/newJob` to `/view/all/newJob` (note the lowercase in the default view name).

jglick commented on core/src/main/java/hudson/model/AllView.java:
```diff
@@ -89,13 +116,70 @@ protected void submit(StaplerRequest req) throws IOException, ServletException,
+            // modern name, we are safe
+            return primaryView;
+        }
+        if (SystemProperties.getBoolean(AllView.class.getName()+".JENKINS-38606", false)) {
```
`getBoolean(String)` suffices.

jglick commented on core/src/main/java/hudson/model/AllView.java:
```diff
@@ -89,13 +116,70 @@ protected void submit(StaplerRequest req) throws IOException, ServletException,
+                    // name conflict, we cannot rename the all view anyway
+                    return primaryView;
+                }
+                if (StringUtils.equals(v.getViewName(), primaryView)) {
```
AFAICT `getViewName()` should be marked `@Nonnull`.

stephenc replied:
Done

jglick replied:
meaning you could use

```java
if (v.getViewName().equals(primaryView)) {
```

jglick commented on core/src/main/java/hudson/model/AllView.java:
```diff
@@ -89,13 +116,70 @@ protected void submit(StaplerRequest req) throws IOException, ServletException,
+                        LOGGER.log(Level.INFO,
+                                "JENKINS-38606 detected for AllView in {0}; renaming view from {1} to {2}",
+                                new Object[]{allView.owner.getUrl(), DEFAULT_VIEW_NAME});
+                        allView.name = DEFAULT_VIEW_NAME;
```
Should this not `save` the owner?

stephenc replied:
nope, responsibility of the caller

jglick commented on core/src/main/java/hudson/model/View.java:
```diff
@@ -1268,6 +1277,9 @@ public static View create(StaplerRequest req, StaplerResponse rsp, ViewGroup own
             if(isXmlSubmission) {
                 View v = createViewFromXML(name, req.getInputStream());
+                if (!owner.getACL().hasCreatePermission(Jenkins.getAuthentication(), owner, v.getDescriptor())) {
+                    throw new Failure(Messages.View_CreatePermissionMissing(v.getDescriptor().getDisplayName()));
```
`AccessDeniedException`, or better yet define a new `ACL.checkCreatePermission` overload for this.

jglick commented on core/src/main/java/hudson/model/View.java:
```diff
@@ -1287,6 +1299,9 @@ public static View create(StaplerRequest req, StaplerResponse rsp, ViewGroup own
             v = descriptor.newInstance(req,req.getSubmittedForm());
         }
+        if (!owner.getACL().hasCreatePermission(Jenkins.getAuthentication(), owner, v.getDescriptor())) {
+            throw new Failure(Messages.View_CreatePermissionMissing(v.getDescriptor().getDisplayName()));
```
ditto

jglick commented on core/src/main/java/hudson/model/ViewDescriptor.java:
```diff
@@ -101,13 +110,65 @@ public AutoCompletionCandidates doAutoCompleteCopyNewItemFrom(@QueryParameter fi
     public List<Descriptor<ListViewColumn>> getColumnsDescriptors() {
-        return ListViewColumn.all();
+        StaplerRequest request = Stapler.getCurrentRequest();
+        View view = request.findAncestorObject(clazz);
```
`request` could be null.

jglick commented on core/src/main/java/hudson/model/ViewDescriptor.java:
```diff
@@ -101,13 +110,65 @@ public AutoCompletionCandidates doAutoCompleteCopyNewItemFrom(@QueryParameter fi
     public List<Descriptor<ViewJobFilter>> getJobFiltersDescriptors() {
-        return ViewJobFilter.all();
+        StaplerRequest request = Stapler.getCurrentRequest();
+        View view = request.findAncestorObject(clazz);
```
ditto

jglick commented on core/src/main/java/hudson/model/ViewDescriptor.java:
```diff
@@ -101,13 +110,65 @@ public AutoCompletionCandidates doAutoCompleteCopyNewItemFrom(@QueryParameter fi
+     * @return the validation result.
+     * @since FIXME
+     */
+    @SuppressWarnings("unused") // expose utility check method to subclasses
```
Is there a particular subclass you had in mind? Not in this PR; some other linked one?

stephenc replied:
Yes but it needs to use the `Referer` header hack to resolve the `View` so probably waiting on dedicated support in stapler for that

jglick replied:
So where is that subclass?

jglick commented on core/src/main/java/hudson/model/ViewDescriptor.java:
```diff
@@ -101,13 +110,65 @@ public AutoCompletionCandidates doAutoCompleteCopyNewItemFrom(@QueryParameter fi
+    /**
+     * Returns true if this {@link View} type is applicable in the specific {@link ViewGroup}.
+     *
+     * @return true to indicate applicable, in which case the view will be instantiable within the given owner.
```
Specify default implementation.

jglick commented on core/src/main/java/hudson/model/ViewDescriptor.java:
```diff
@@ -101,13 +110,65 @@ public AutoCompletionCandidates doAutoCompleteCopyNewItemFrom(@QueryParameter fi
+    /**
+     * Returns true if this {@link View} type is applicable to the given {@link ViewGroup} type.
+     *
+     * @return true to indicate applicable, in which case the view will be instantiable within the type of owner.
```
Specify default implementation.

jglick commented on core/src/main/resources/hudson/model/MyViewsProperty/newView.jelly:
```diff
@@ -32,7 +32,7 @@ THE SOFTWARE.
 
     <l:main-panel>
-      <j:invokeStatic var="views" className="hudson.model.View" method="allInstantiable" />
+      <j:invokeStatic var="views" className="hudson.model.View" method="allInstantiable"/>
```
Gratuitous diff hunk.

jglick commented on core/src/main/resources/jenkins/model/Jenkins/newView.jelly:
```diff
@@ -31,7 +31,7 @@ THE SOFTWARE.
 
     <l:main-panel>
-      <j:invokeStatic var="views" className="hudson.model.View" method="allInstantiable" />
+      <j:invokeStatic var="views" className="hudson.model.View" method="allInstantiable"/>
```
ditto

jglick reviewed (changes requested):
Some minor inline comments, and [tests are failing](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/6417/testReport/); :+1: on concept.

stephenc committed:
[JENKINS-39300] Address review comments

stephenc reviewed (commented):
@jglick I think i have addressed all your comments (in some way or other)

jglick commented on core/src/main/java/hudson/views/ListViewColumn.java:
```diff
@@ -121,20 +122,50 @@ public boolean shownByDefault() {
-                if (!lvc.shownByDefault())      continue; // skip this
+                if (!lvc.shownByDefault()) {
+                    continue; // skip this
+                }
```
gratuitous reformatting

jglick commented on core/src/main/java/hudson/model/AllView.java:
```diff
@@ -89,13 +116,72 @@ protected void submit(StaplerRequest req) throws IOException, ServletException,
+            // modern name, we are safe
+            return primaryView;
+        }
+        if (SystemProperties.getBoolean(AllView.class.getName()+".JENKINS-38606")) {
```
Making this disabled by default makes it less than useful.

IIUC the concern is about compatibility for direct links to, say, `http://jenkins/view/All/job/something/`. But that is easily solved: amend `ViewGroupMixIn.getView` (called from `Jenkins.getView(String)`, the Stapler URL binding) to fall back to searching by `View.getDisplayName` if no match is found via `getViewName`.

jglick reviewed (approved):
Would like to discuss having better URL compatibility, and would like to see the downstream PR that various API methods here imply.

stephenc committed:
[JENKINS-38606] Allow legacy localized links to continue working after migration

- This means we can default the migration to enabled

stephenc commented:
See https://github.com/jenkinsci/cloudbees-folder-plugin/pull/79 for example usage of `migrateLegacyPrimaryAllViewLocalizedName` method

stephenc commented:
@rsandell @daniel-beck @jglick PR has been updated

jglick commented on core/src/main/java/hudson/model/ViewGroupMixIn.java:
```diff
@@ -99,6 +101,15 @@ public View getView(String name) {
             View pv = getPrimaryView();
             if (pv instanceof ViewGroup)
                 return ((ViewGroup)pv).getView(name);
+            if (pv instanceof AllView && AllView.DEFAULT_VIEW_NAME.equals(pv.name)) {
```
Sounds right but hard for me at least to follow how this will work in progress; recommend having some code coverage using `@LocalData` to prove it works.

jglick commented on core/src/main/java/hudson/security/ACL.java:
```diff
@@ -120,6 +123,42 @@ public boolean hasCreatePermission(@Nonnull Authentication a, @Nonnull ItemGroup
+     * @param c the container of the item.
+     * @param d the descriptor of the view to be created.
+     * @throws AccessDeniedException if the user doesn't have the permission.
+     * @since 1.607
```
`FIXME`

jglick reviewed (approved):
Sounds good to the extent I understand it.

jglick reviewed (changes requested):
Ah but lots of test failures.

stephenc committed:
[JENKINS-39300] Update tests

daniel-beck commented:
One test still fails:

	Expected: a string containing "ERROR: Jenkins does not allow to delete 'All' view"
	     but: was "
	ERROR: Jenkins does not allow to delete 'all' view
	"

daniel-beck commented on core/src/main/java/hudson/model/ViewGroupMixIn.java:
```diff
@@ -99,6 +101,15 @@ public View getView(String name) {
+            if (pv instanceof AllView && AllView.DEFAULT_VIEW_NAME.equals(pv.name)) {
+                // JENKINS-38606: primary view is the default AllView, is somebody using an old link to localized form?
+                for (Locale l : Locale.getAvailableLocales()) {
+                    if (name.equals(Messages._Hudson_ViewName().toString(l))) {
```
So changing the translations will break this compatibility measure for that language?

stephenc committed:
[JENKINS-39300] Fix final test case

stephenc commented:
I think we're good to merge now... i'll hold off until tomorrow and then merge just in case anyone has and issues with the test changes

jglick commented on core/src/main/java/hudson/model/AllView.java:
```diff
@@ -89,13 +116,73 @@ protected void submit(StaplerRequest req) throws IOException, ServletException,
+                        // bingo JENKINS-38606 detected
+                        LOGGER.log(Level.INFO,
+                                "JENKINS-38606 detected for AllView in {0}; renaming view from {1} to {2}",
+                                new Object[]{allView.owner.getUrl(), DEFAULT_VIEW_NAME});
```
#3088
